core/state_transition.go

```diff
@@ -395,45 +388,47 @@ func (st *StateTransition) distributeTxFees() error {
 	refund := new(big.Int).Mul(new(big.Int).SetUint64(st.gas), st.gasPrice)
 	gasUsed := new(big.Int).SetUint64(st.gasUsed())
 	totalTxFee := new(big.Int).Mul(gasUsed, st.gasPrice)
+	from := st.msg.From()
 
 	// Divide the transaction into a base (the minimum transaction fee) and tip (any extra).
 	baseTxFee := new(big.Int).Mul(gasUsed, st.gasPriceMinimum)
 	tipTxFee := new(big.Int).Sub(totalTxFee, baseTxFee)
+	feeCurrency := st.msg.FeeCurrency()
 
-	// Pay gateway fee to the specified recipient.
-	if st.msg.GatewayFeeRecipient() != nil {
-		log.Trace("Crediting gateway fee", "recipient", *st.msg.GatewayFeeRecipient(), "amount", st.msg.GatewayFee(), "feeCurrency", st.msg.FeeCurrency())
-		if err := st.creditFee(*st.msg.GatewayFeeRecipient(), st.msg.GatewayFee(), st.msg.FeeCurrency()); err != nil {
-			log.Error("Failed to credit gateway fee", "err", err)
-			return err
-		}
+	gatewayFeeRecipient := st.msg.GatewayFeeRecipient()
+	if gatewayFeeRecipient == nil {
+		gatewayFeeRecipient = &common.ZeroAddress
 	}
 
-	log.Trace("Crediting gas fee tip", "recipient", st.evm.Coinbase, "amount", tipTxFee, "feeCurrency", st.msg.FeeCurrency())
-	if err := st.creditFee(st.evm.Coinbase, tipTxFee, st.msg.FeeCurrency()); err != nil {
+	governanceAddress, err := vm.GetRegisteredAddressWithEvm(params.GovernanceRegistryId, st.evm)
+	if err != nil && err != commerrs.ErrSmartContractNotDeployed && err != commerrs.ErrRegistryContractNotDeployed {
 		return err
+	} else if err != nil {
+		log.Trace("Cannot credit gas fee to community fund: refunding fee to sender", "error", err, "fee", baseTxFee)
+		governanceAddress = &common.ZeroAddress
+		refund.Add(refund, baseTxFee)
+		baseTxFee = new(big.Int)
 	}
 
-	// Send the base of the transaction fee to the community fund.
-	governanceAddress, err := vm.GetRegisteredAddressWithEvm(params.GovernanceRegistryId, st.evm)
-	if err != nil {
-		if err != commerrs.ErrSmartContractNotDeployed && err != commerrs.ErrRegistryContractNotDeployed {
-			return err
+	log.Trace("distributeTxFees", "from", from, "refund", refund, "feeCurrency", st.msg.FeeCurrency(),
+		"gatewayFeeRecipient", *gatewayFeeRecipient, "gatewayFee", st.msg.GatewayFee(),
+		"coinbaseFeeRecipient", st.evm.Coinbase, "coinbaseFee", tipTxFee,
+		"comunityFundRecipient", *governanceAddress, "communityFundFee", baseTxFee)
+	if feeCurrency == nil {
+		if gatewayFeeRecipient != &common.ZeroAddress {
```

Comment:
Should we refund the gateway fee in this case?

Reply:
No, since on `payFees` we don't debit it if `st.msg` doesn't have `GatewayFeeRecipient`